### ISSUE TITLE
Occasionally the city view masks out the incorrect tiles

### ISSUE THREAD

**Issue by zxcvforz** [chaorace/cqui#45]
![289070_20161106082537_1](https://cloud.githubusercontent.com/assets/6653593/20034505/c9c5f788-a3fa-11e6-89af-9ac2ce75b845.jpg)

This is the same problem as in the vanilla game when we enable both the tile-purchase mode and the citizen management mode. UI functions as normal, able to manage citizens and buy tiles. The issue is only cosmetic.

**Comment by chaorace** [comment]
@zxcvforz Could you provide the affected savefile? Thanks!

**Comment by zxcvforz** [comment]
Here it is. My apologies for the inconvenience. Should require the Aztec DLC.

[CQUI_Issue_45_CityViewMasking.zip](https://github.com/chaorace/cqui/files/573582/CQUI_Issue_45_CityViewMasking.zip)

**Comment by zxcvforz** [comment]
I've been testing this and realized that I can only reproduce it as Russia. Not only that, some Russian cities don't show the masking error, some do. Same as in vanilla. The number of tiles properly unmasked also isn't the same as the extra tiles Russia gets.

To clarify, in vanilla, start as Russia, build city, then select city. Now turn on *both* the tile purchase mode and citizen management mode, and the same masking error shows. Doing the same as other civs don't show a problem. The bug is clearly coming from something already in game, and somehow associated with Russia.

Here's a new save. The three cities on the eastern end of the empire don't show the masking error, while captured enemy capitals (Kyoto and Achen) do.

[CQUI_Issue_45_RussiaMaskingErrorNotAlwaysShown.zip](https://github.com/chaorace/cqui/files/576797/CQUI_Issue_45_RussiaMaskingErrorNotAlwaysShown.zip)

I'm going to keep investigating this as Russia. Russia is my personal Civ6 boom at the moment. :dancing_men:

**Comment by chaorace** [comment]
I get the feeling that this may be most easily fixed by gutting out the individual masking calls from the tile purchase / citizen management views and consolidating them in the same place the cityview toggling is handled. Let's see if this is an easy fix, eh?

**Comment by zxcvforz** [comment]
Might be. I put in some print statements in the PlotInfo::AggregateLensHexes() function and found that the m_kLensMask[PLOT_PURCHASE] table has some tiles listed twice, but the m_kLensMask[CITIZEN_MANAGEMENT] table is always just listing the right amount of tiles.

But then even with cities that show the correct masking, there can be repetitions in the PLOT_PURCHASE table, so I don't know what's happening there.

Perhaps pruning the lens mask table while in AggregateLensHexes() function to remove repetition could help? I'm taking wild guesses now, though.

**Comment by jacks0nX** [comment]
Perhaps some internal error due to Russias ability to grab more land upon founding a city? Doesn't really help us, but maybe worth a bug report to firaxis.

**Comment by zxcvforz** [comment]
Finally I caught the moment when the masking error occurs! And I have no idea why it does this.

In the attached save, London is building a Shrine, and no cities yet show the masking error. But end the turn and London starts getting the masking error. This clearly is not limited to Russia's extra land ability.

And I also went back to check the previous two saves that I uploaded, neither the Shrine nor the Holy Site is the culprit; in my earliest upload above, one of the cities with the masking error doesn't even have any districts built, nor did it come in range of any other city's districts.

There must be something more fundamental to this masking error.

[EndTurnAndLondonGetsMaskingError.zip](https://github.com/chaorace/cqui/files/585202/EndTurnAndLondonGetsMaskingError.zip)
